### PR TITLE
Remove powers-of-2 restriction on cohort size

### LINKED ISSUE #130: Remove powers-of-2 restriction on cohort size
Given that our cohorts will be likely small (i.e. less than 100), we're not going to take advantage of FFT optimizations that introduce the requirement for the cohort size to be a power of 2.

### PR DESCRIPTION
**Pull request by piotr-roslaniec**
**Type of PR:**
- Feature

**Required reviews:** 
- 2

**What this does:**
- Adds benchmarks to compare the performance of `MixedRadixEvaluationDomain` (without constraints on the number of coefficients) and `Radix2EvaluationDomain` (number of coefficients must be a power of two)
- Replaces usage of `Radix2EvaluationDomain` with [`GeneralEvaluationDomain`](https://docs.rs/ark-poly/latest/ark_poly/domain/general/enum.GeneralEvaluationDomain.html#) (automatically select domain size) in `ferveo`

**Issues fixed/closed:**
- Fixes #130

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
- Please see benchmarking results attached
- Based on https://github.com/nucypher/ferveo/pull/136
- Fixed a bug where `Dkg` in `ferveo::api` is using an incorrect number of domain points to create a precomputed share

<details>
<summary><b>Benchmarking results</b></summary>

lines
violin
</details>

### CODE REVIEW

**Review by cygnusv — APPROVED** [review]
I thought this would be more painful! Great work @piotr-roslaniec, as usual 💪

File: ferveo/src/dkg.rs [hunk 1]
```diff
@@ -78,16 +76,11 @@ impl<E: Pairing> PubliclyVerifiableDkg<E> {
...
-            return Err(Error::InvalidShareNumberParameter(
-                dkg_params.shares_num,
-            ));
-        }
-        let domain = ark_poly::Radix2EvaluationDomain::<E::ScalarField>::new(
-            dkg_params.shares_num as usize,
-        )
-        .expect("unable to construct domain");
+        let domain =
+            ark_poly::MixedRadixEvaluationDomain::<E::ScalarField>::new(
```
Comment (derekpierre) [inline ferveo/src/dkg.rs:80]: Nice! That's so clean.

File: ferveo/benches/benchmarks/eval_domain.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+const NUM_SHARES_CASES: [usize; 6] = [2, 4, 8, 16, 32, 64];
+
+pub fn bench_eval_domain(c: &mut Criterion) {
+    let mut group = c.benchmark_group("EVAL DOMAIN");
+    group.sample_size(10);
+
+    let rng = &mut StdRng::seed_from_u64(0);
+    let s = ark_bls12_381::Fr::from_random_bytes(&[0u8; 32]).unwrap();
+
+    for shares_num in NUM_SHARES_CASES {
```
Comment (derekpierre) [inline ferveo/benches/benchmarks/eval_domain.rs:21]: So this is benchmarking the speed of num shares that are powers of 2 for Radix2 vs MixedRadix? Cool.

As an aside. Do we have a test somewhere for a non-power of 2 number of shares? Basically, a test that confirms that the non-powers of 2 work appropriately.

Comment (piotr-roslaniec) [reply ferveo/benches/benchmarks/eval_domain.rs:21]: I didn't consider this kind of test to be interesting before. I did a manual test (in REPL) before submitting the PR and that lulled me into a false sense of correctness. Now that I followed up on your suggestion, it turns out only the simple variant works for a non-power-of-two number of shares. 

I will post the changes here and investigate the issue with the precomputed variant.

Thanks for catching this.

File: ferveo/src/api.rs [hunk 3]
```diff
@@ -374,179 +374,191 @@ mod test_ferveo_api {
...
-
-        let shared_secret = share_combine_precomputed(&decryption_shares);
-        let result = decrypt_with_shared_secret(
-            &ciphertext,
-            aad,
-            &SharedSecret(shared_secret),
-        );
-        assert!(result.is_err());
+        // Works for both power of 2 and non-power of 2
+        for shares_num in [4, 7] {
```
Comment (derekpierre) [inline ferveo/src/api.rs:378]: Good to see - 🔥

File: ferveo/src/api.rs [hunk 4]
```diff
@@ -238,7 +238,12 @@ impl AggregatedTranscript {
         aad: &[u8],
         validator_keypair: &Keypair,
     ) -> Result<DecryptionSharePrecomputed> {
-        let domain_points: Vec<_> = dkg.0.domain.elements().collect();
+        let domain_points: Vec<_> = dkg
+            .0
+            .domain
+            .elements()
+            .take(dkg.0.dkg_params.shares_num as usize)
+            .collect();
```
Comment (piotr-roslaniec) [inline ferveo/src/api.rs:246]: This is the root cause of the issue found by unit tests. The domain may have an arbitrary size, but we must use at most `shares_num` domain points.

File: ferveo/benches/benchmarks/eval_domain.rs (hunk 2 above)
Comment (piotr-roslaniec) [reply ferveo/benches/benchmarks/eval_domain.rs:21]: The bug has been found and fixed. Please see my comment below.

**Review by derekpierre — APPROVED** [review]
🎸

**Pull request merged**